quinn-proto/src/endpoint.rs

```diff
@@ -494,22 +494,20 @@ impl Endpoint {
 
         let server_config = self.server_config.as_ref().unwrap().clone();
 
-        let (retry_src_cid, orig_dst_cid) = if header.token.is_empty() {
-            (None, header.dst_cid)
+        let token_state = if header.token.is_empty() {
+            IncomingTokenState::default(&header)
         } else {
-            match RetryToken::decode(
+            let token = RetryToken::decode(
                 &*server_config.token_key,
                 &addresses.remote,
                 &header.dst_cid,
                 &header.token,
-            ) {
-                Ok(token)
-                    if token.issued + server_config.retry_token_lifetime > SystemTime::now() =>
-                {
-                    (Some(header.dst_cid), token.orig_dst_cid)
-                }
-                Err(ValidationError::Ignore) => (None, header.dst_cid),
-                _ => {
+            );
+            let token_state = token.and_then(|token| token.validate(&header, &server_config));
```

Comment:
I like this direction, why not fold `decode()` and `validate()` into a single method?

Reply:
I think that there's comprehensibility benefits to our code having a separation between the phases of 1. decrypting the encrypted payload 2. deserializating the decrypted payload 3. validating the deserialized payload's applicability to the current connection. If we wanted to go the extreme opposite direction, we could remove `Token` and its contents as structs altogether and purely have it be a bunch of functions that call each other without ever reifying their intermediate states as structs. I think that would be worse--although it would probably be fewer lines of code.

I just pushed a commit b4a4f87caed4d7659bfe2ba2e59e42dee516ec34 which moves the calls to `decode()` and to `validate()` into a new `IncomingToken.handle_header` method. I think it's kinda nice, as it moves even more logic from `endpoint.rs` into `token.rs`. Let me know what you think of the idea?

`IncomingToken` ends up with this method:

```rs
/// Construct for an `Incoming` given the first packet header, or error if the connection
/// cannot be established
pub(crate) fn handle_header(
    header: &InitialHeader,
    server_config: &ServerConfig,
    address: SocketAddr,
) -> Result<Self, InvalidRetryTokenError> {
    if header.token.is_empty() {
        return Ok(Self::default(&header));
    }

    Token::decode(&*server_config.token_key, &header.token)
        .and_then(|token| token.validate(&header, &server_config, address))
        .or_else(|e| match e {
            ValidationError::Ignore => Ok(Self::default(&header)),
            ValidationError::InvalidRetry => Err(InvalidRetryTokenError),
        })
}
```

Which leaves the corresponding logic in `Endpoint.handle_first_packet` as basically just calling it and the initial close case:

```rs
let token = match IncomingToken::handle_header(&header, &server_config, addresses.remote) {
    Ok(token) => token,
    Err(InvalidRetryTokenError) => {
        debug!("rejecting invalid retry token");
        return Some(DatagramEvent::Response(self.initial_close(
            header.version,
            addresses,
            &crypto,
            &header.src_cid,
            TransportError::INVALID_TOKEN(""),
            buf,
        )));
    }
};
```

Reply:
Yes, that seems like a nice improvement! I would probably rename `handle_header()` to `from_header()` and it potentially still seems interesting to me to inline the logic of `validate()` (assuming it has no other significant callers).